fraud/service.go

```diff
@@ -0,0 +1,147 @@
+package fraud
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"sync"
+	"time"
+
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+)
+
+// service is responsible for validating and propagating Fraud Proofs.
+// It implements the Service interface.
+type service struct {
+	mu           sync.RWMutex
+	unmarshalers map[ProofType]ProofUnmarshaler
+
+	topics *topics
+	getter headerFetcher
+}
+
+func NewService(p *pubsub.PubSub, getter headerFetcher) Service {
+	return &service{
+		topics: &topics{
+			pubsub:       p,
+			pubSubTopics: make(map[ProofType]*pubsub.Topic),
+		},
+		unmarshalers: make(map[ProofType]ProofUnmarshaler),
+		getter:       getter,
+	}
+}
+
+func (f *service) Subscribe(proofType ProofType) (Subscription, error) {
+	// TODO: @vgonkivs check if fraud proof is in fraud store, then return with error
+	f.mu.Lock()
```

Comment:
pls defer f.mu.Unlock here bc if !ok and it returns, map is still locked.

Reply:
Also if you unlock at the end of the func, might as well be a defer instead of 3 diff unlock statements